pkg/cloud/baremetal/actuators/machine/actuator.go

```diff
@@ -963,7 +973,106 @@ func (a *Actuator) remediateIfNeeded(ctx context.Context, machine *machinev1.Mac
 		return &clustererror.RequeueAfterError{RequeueAfter: time.Second * 5}
 	}
 
-	// node is now available again
+	// node is now available again - restore annotations and labels
+	if _, annotationsBackupExists := machine.Annotations[nodeAnnotationsBackupAnnotation]; annotationsBackupExists {
+		return a.restoreAnnotationsAndLabels(node, machine, ctx)
+	}
+
+	//remediation is done
 	log.Printf("Node %s is available, remediation of Machine %s complete", node.Name, machine.Name)
 	return a.deleteRemediationAnnotations(ctx, machine)
 }
+
+// storeAnnotationsAndLabels copies node's annotations and labels and stores them on machine's annotations
+func (a *Actuator) storeAnnotationsAndLabels(node *corev1.Node, machine *machinev1.Machine, ctx context.Context) error {
+	marshaledAnnotations, err := marshal(node.Annotations)
+	if err != nil {
+		log.Printf("Failes to marshal node %s annotations assoicated with Machine %s: %s",
+			node.Name, machine.Name, err.Error())
+		return err
+	}
+
+	marshaledLabels, err := marshal(node.Labels)
+	if err != nil {
+		log.Printf("Failes to marshal node %s labels assoicated with Machine %s: %s",
+			node.Name, machine.Name, err.Error())
+		return err
+	}
+
+	machine.Annotations[nodeLabelsBackupAnnotation] = marshaledLabels
+	machine.Annotations[nodeAnnotationsBackupAnnotation] = marshaledAnnotations
+
+	err = a.client.Update(ctx, machine)
+	if err != nil {
+		log.Printf("failed to update machine with node's annotations and labels %s: %s", machine.Name, err.Error())
+		return err
+	}
+
+	return &clustererror.RequeueAfterError{}
+}
+
+// restoreAnnotationsAndLabels copies annotations and labels stored in machine's annotation to the node
+// and removes the annotations used to store that data
+func (a *Actuator) restoreAnnotationsAndLabels(node *corev1.Node, machine *machinev1.Machine, ctx context.Context) error {
+	if len(machine.Annotations) == 0 {
+		return &clustererror.RequeueAfterError{}
+	}
+
+	nodeAnn, err := unmarshal(machine.Annotations[nodeAnnotationsBackupAnnotation])
+	if err != nil {
+		log.Printf("failed to unmarshal node's annotations from %s: %s", machine.Name, err.Error())
+		return err
```

Comment:
I think we have to be more pessimistic here.
If our stored annotations are corrupted or removed, then I don't think we can error out without trying the labels too.

Reply:
If there's a corrupted value, we will return err forever. Only after we restore annotations, we try to restore labels. 

Do you think we should try to restore annotations and if it doesn't succeed, just skip it?